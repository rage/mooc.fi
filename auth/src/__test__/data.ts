import { NextFunction, Request, Response } from "express"

export const testProfile = {
  edu_person_affiliation: "",
  edu_person_principal_name: "test",
  schac_home_organization: "helsinki.fi",
  organizational_unit: "organizational_unit",
  schac_personal_unique_code:
    "urn:schac:personalUniqueCode:int:studentID:helsinki.fi:121345678",
  display_name: "test",
  given_name: "test",
  mail: "test@mail.com",
}

export const req = {
  body: {},
  query: {
    RelayState: JSON.stringify({
      language: "fi",
      provider: "hy",
      action: "sign-in",
    }),
  },
  params: {},
  login: jest.fn().mockImplementation((_: any, fn: Function) => fn(null)),
  logout: jest.fn(),
} as unknown as Request
export const res = {
  redirect: jest.fn().mockReturnThis(),
  locals: {},
  setMOOCCookies: jest.fn().mockReturnThis(),
} as unknown as Response
export const next = jest.fn() as NextFunction

export const okTokenResponse = {
  data: {
    status: 200,
    success: true,
    tmc_token: "tmc_token",
    access_token: "access_token",
    admin: false,
  },
}
