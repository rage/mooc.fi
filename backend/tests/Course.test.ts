import { gql } from "graphql-request"
import { getTestContext, fakeTMC } from "./__helpers"
import {
  //normalUser,
  normalUserDetails,
  //adminUser,
  adminUserDetails,
} from "./data"
import { seed } from "./data/seed"

import { Course } from "@prisma/client"
import { orderBy } from "lodash"

const ctx = getTestContext()
const tmc = fakeTMC({
  "Bearer normal": [200, normalUserDetails],
  "Bearer admin": [200, adminUserDetails],
})

const courseQuery = gql`
  query course($id: ID, $slug: String, $language: String) {
    course(id: $id, slug: $slug, language: $language) {
      id
      name
      slug
      description
      link
    }
  }
`

const fullCourseQuery = gql`
  query course($id: ID, $slug: String, $language: String) {
    course(id: $id, slug: $slug, language: $language) {
      id
      name
      slug
      ects
      order
      study_module_order
      teacher_in_charge_name
      teacher_in_charge_email
      support_email
      start_date
      end_date
      tier
      photo {
        id
        compressed
        compressed_mimetype
        uncompressed
        uncompressed_mimetype
      }
      promote
      start_point
      hidden
      study_module_start_point
      status
      course_translations {
        id
        name
        language
        description
        link
      }
      open_university_registration_links {
        id
        course_code
        language
        link
      }
      study_modules {
        id
      }
      course_variants {
        id
        slug
        description
      }
      course_aliases {
        id
        course_code
      }
      inherit_settings_from {
        id
      }
      completions_handled_by {
        id
      }
      has_certificate
      user_course_settings_visibilities {
        id
        language
      }
      upcoming_active_link
      automatic_completions
      automatic_completions_eligible_for_ects
      exercise_completions_needed
      points_needed
    }
  }
`

const coursesQuery = gql`
  query AllCourses($language: String, $orderBy: CourseOrderByInput) {
    courses(orderBy: $orderBy, language: $language) {
      id
      slug
      name
      order
      study_module_order
      photo {
        id
        compressed
        uncompressed
      }
      promote
      status
      start_point
      study_module_start_point
      hidden
      description
      link
      upcoming_active_link
      study_modules {
        id
        slug
      }
      course_translations {
        id
        language
        name
      }
      user_course_settings_visibilities {
        id
        language
      }
    }
  }
`

const courseExistsQuery = gql`
  query courseExists($slug: String!) {
    course_exists(slug: $slug)
  }
`

const createCourseMutation = gql`
  mutation createCourse($course: CourseCreateArg!) {
    addCourse(course: $course) {
      id
      name
      slug
      ects
      order
      study_module_order
      teacher_in_charge_name
      teacher_in_charge_email
      support_email
      start_date
      end_date
      tier
      photo {
        id
        compressed
        compressed_mimetype
        uncompressed
        uncompressed_mimetype
      }
      promote
      start_point
      hidden
      study_module_start_point
      status
      course_translations {
        id
        name
        language
        description
        link
      }
      open_university_registration_links {
        id
        course_code
        language
        link
      }
      study_modules {
        id
      }
      course_variants {
        id
        slug
        description
      }
      course_aliases {
        id
        course_code
      }
      inherit_settings_from {
        id
      }
      completions_handled_by {
        id
      }
      has_certificate
      user_course_settings_visibilities {
        id
        language
      }
      upcoming_active_link
      automatic_completions
      automatic_completions_eligible_for_ects
      exercise_completions_needed
      points_needed
    }
  }
`

// study_modules may be returned in any order, let's just sort them so snapshots are equal
const sortStudyModules = (course: any) => {
  if (!course.study_modules) {
    return course
  }

  return {
    ...course,
    study_modules: orderBy(course.study_modules, ["id"], ["asc"]),
  }
}

describe("course queries", () => {
  beforeAll(() => tmc.setup())
  afterAll(() => tmc.teardown())

  describe("course", () => {
    let createdCourses: Course[] | null = null

    beforeEach(async () => {
      createdCourses = (await seed(ctx.prisma)).courses
    })

    afterEach(() => (createdCourses = null))

    describe("normal user", () => {
      beforeEach(() => ctx!.client.setHeader("Authorization", "Bearer normal"))

      it("should error on no parameters", async () => {
        try {
          await ctx.client.request(courseQuery)
          fail()
        } catch {}
      })

      it("returns course on id and slug", async () => {
        const resId = await ctx.client.request(courseQuery, {
          id: createdCourses![0].id,
        })
        const resSlug = await ctx.client.request(courseQuery, {
          slug: "course1",
        })

        ;[resId, resSlug].map((res) =>
          expect(sortStudyModules(res.course)).toMatchSnapshot({
            id: expect.any(String),
          }),
        )
      })

      it("returns correct language", async () => {
        const res = await ctx.client.request(courseQuery, {
          slug: "course1",
          language: "en_US",
        })

        expect(sortStudyModules(res.course)).toMatchSnapshot({
          id: expect.any(String),
        })
      })

      it("should return null on non-existent language", async () => {
        const res = await ctx.client.request(courseQuery, {
          slug: "course1",
          language: "sv_SE",
        })

        expect(res).toEqual({ course: null })
      })

      it("should error on invalid id and slug", async () => {
        try {
          await ctx.client.request(courseQuery, {
            id: new Array(33).join("1"),
          })
          fail()
        } catch {}
        try {
          await ctx.client.request(courseQuery, {
            slug: "invalid",
          })
          fail()
        } catch {}
      })
    })

    describe("admin", () => {
      beforeEach(() => ctx!.client.setHeader("Authorization", "Bearer admin"))

      it("returns full course on id and slug", async () => {
        const resId = await ctx.client.request(fullCourseQuery, {
          id: createdCourses![0].id,
        })
        const resSlug = await ctx.client.request(fullCourseQuery, {
          slug: "course1",
        })

        ;[resId, resSlug].map((res) =>
          expect(sortStudyModules(res.course)).toMatchSnapshot({
            id: expect.any(String),
          }),
        )
      })
    })
  })

  describe("courses", () => {
    beforeEach(async () => {
      await seed(ctx.prisma)
    })

    it("returns courses", async () => {
      const res = await ctx.client.request(coursesQuery)

      expect(
        orderBy(res.courses.map(sortStudyModules), ["id"]),
      ).toMatchSnapshot()
    })

    it("returns courses ordered", async () => {
      await Promise.all(
        ["asc", "desc"].map(async (order: string) => {
          const res = await ctx.client.request(coursesQuery, {
            orderBy: { name: order },
          })

          expect(res.courses.map(sortStudyModules)).toMatchSnapshot(
            `courses-${order}`,
          )

          return null
        }),
      )
    })

    it("returns courses filtered by language", async () => {
      await Promise.all(
        ["fi_FI", "en_US", "bogus"].map(async (language: string) => {
          const res = await ctx.client.request(coursesQuery, {
            language,
          })

          expect(
            orderBy(res.courses?.map(sortStudyModules), ["id"]),
          ).toMatchSnapshot(`courses-${language}`)

          return null
        }),
      )
    })
  })

  describe("course_exists", () => {
    beforeEach(async () => {
      await seed(ctx.prisma)
      ctx!.client.setHeader("Authorization", "Bearer normal")
    })

    it("returns true on existing course", async () => {
      const res = await ctx.client.request(courseExistsQuery, {
        slug: "course1",
      })

      expect(res).toEqual({ course_exists: true })
    })

    it("returns false on non-existing course", async () => {
      const res = await ctx.client.request(courseExistsQuery, {
        slug: "bogus",
      })

      expect(res).toEqual({ course_exists: false })
    })
  })
})

describe("course mutations", () => {
  beforeAll(() => tmc.setup())
  afterAll(() => tmc.teardown())

  const newCourse = {
    name: "new1",
    slug: "new1",
    start_date: "01/01/1900",
    teacher_in_charge_email: "e@mail.com",
    teacher_in_charge_name: "teacher",
    study_modules: [{ id: "00000000000000000000000000000101" }],
    course_translations: [
      {
        description: "description_en_US",
        language: "en_US",
        name: "name_en_US",
      },
    ],
  }

  describe("addCourse", () => {
    beforeEach(async () => {
      await seed(ctx.prisma)
      ctx!.client.setHeader("Authorization", "Bearer admin")
    })

    it("creates a course ", async () => {
      const res = await ctx!.client.request(createCourseMutation, {
        course: newCourse,
      })

      expect(res).toMatchSnapshot({
        addCourse: {
          id: expect.any(String),
          course_translations: [
            {
              id: expect.any(String),
            },
          ],
        },
      })

      const createdCourse = await ctx.prisma.course.findFirst({
        where: { slug: "new1" },
      })

      expect(createdCourse).not.toEqual(null)
      expect(createdCourse!.id).toEqual(res.addCourse.id)
      expect(createdCourse).toMatchSnapshot({
        created_at: expect.any(Date),
        updated_at: expect.any(Date),
        id: expect.any(String),
      })
    })
  })
})
