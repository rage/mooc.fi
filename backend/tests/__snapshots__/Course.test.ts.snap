// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`course mutations addCourse creates a course  1`] = `
Object {
  "addCourse": Object {
    "automatic_completions": false,
    "automatic_completions_eligible_for_ects": true,
    "completions_handled_by": null,
    "course_aliases": Array [],
    "course_translations": Array [
      Object {
        "description": "description_en_US",
        "id": Any<String>,
        "language": "en_US",
        "link": null,
        "name": "name_en_US",
      },
    ],
    "course_variants": Array [],
    "ects": null,
    "end_date": null,
    "exercise_completions_needed": null,
    "has_certificate": false,
    "hidden": null,
    "id": Any<String>,
    "inherit_settings_from": null,
    "name": "new1",
    "open_university_registration_links": Array [],
    "order": null,
    "photo": null,
    "points_needed": null,
    "promote": null,
    "slug": "new1",
    "start_date": "01/01/1900",
    "start_point": null,
    "status": "Upcoming",
    "study_module_order": null,
    "study_module_start_point": null,
    "study_modules": Array [
      Object {
        "id": "00000000-0000-0000-0000-000000000101",
      },
    ],
    "support_email": null,
    "teacher_in_charge_email": "e@mail.com",
    "teacher_in_charge_name": "teacher",
    "tier": null,
    "upcoming_active_link": false,
    "user_course_settings_visibilities": Array [],
  },
}
`;

exports[`course mutations addCourse creates a course  2`] = `
Object {
  "automatic_completions": false,
  "automatic_completions_eligible_for_ects": true,
  "completion_email_id": null,
  "completions_handled_by_id": null,
  "created_at": Any<Date>,
  "ects": null,
  "end_date": null,
  "exercise_completions_needed": null,
  "has_certificate": false,
  "hidden": null,
  "id": Any<String>,
  "inherit_settings_from_id": null,
  "name": "new1",
  "order": null,
  "owner_organization_id": null,
  "photo_id": null,
  "points_needed": null,
  "promote": null,
  "slug": "new1",
  "start_date": "01/01/1900",
  "start_point": null,
  "status": "Upcoming",
  "study_module_order": null,
  "study_module_start_point": null,
  "support_email": null,
  "teacher_in_charge_email": "e@mail.com",
  "teacher_in_charge_name": "teacher",
  "tier": null,
  "upcoming_active_link": false,
  "updated_at": Any<Date>,
}
`;

exports[`course queries course admin returns full course on id and slug 1`] = `
Object {
  "automatic_completions": false,
  "automatic_completions_eligible_for_ects": true,
  "completions_handled_by": null,
  "course_aliases": Array [],
  "course_translations": Array [
    Object {
      "description": "course1_description_en_US",
      "id": "00000000-0000-0000-0000-000000000011",
      "language": "en_US",
      "link": "http://link.com",
      "name": "course1_en_US",
    },
    Object {
      "description": "course1_description_fi_FI",
      "id": "00000000-0000-0000-0000-000000000012",
      "language": "fi_FI",
      "link": "http:/link.fi.com",
      "name": "course1_fi_FI",
    },
  ],
  "course_variants": Array [],
  "ects": null,
  "end_date": "12/31/2100",
  "exercise_completions_needed": null,
  "has_certificate": false,
  "hidden": null,
  "id": Any<String>,
  "inherit_settings_from": null,
  "name": "course1",
  "open_university_registration_links": Array [],
  "order": null,
  "photo": null,
  "points_needed": null,
  "promote": null,
  "slug": "course1",
  "start_date": "01/01/1900",
  "start_point": null,
  "status": "Upcoming",
  "study_module_order": null,
  "study_module_start_point": null,
  "study_modules": Array [
    Object {
      "id": "00000000-0000-0000-0000-000000000101",
    },
    Object {
      "id": "00000000-0000-0000-0000-000000000102",
    },
  ],
  "support_email": null,
  "teacher_in_charge_email": "e@mail.com",
  "teacher_in_charge_name": "teacher",
  "tier": null,
  "upcoming_active_link": false,
  "user_course_settings_visibilities": Array [],
}
`;

exports[`course queries course admin returns full course on id and slug 2`] = `
Object {
  "automatic_completions": false,
  "automatic_completions_eligible_for_ects": true,
  "completions_handled_by": null,
  "course_aliases": Array [],
  "course_translations": Array [
    Object {
      "description": "course1_description_en_US",
      "id": "00000000-0000-0000-0000-000000000011",
      "language": "en_US",
      "link": "http://link.com",
      "name": "course1_en_US",
    },
    Object {
      "description": "course1_description_fi_FI",
      "id": "00000000-0000-0000-0000-000000000012",
      "language": "fi_FI",
      "link": "http:/link.fi.com",
      "name": "course1_fi_FI",
    },
  ],
  "course_variants": Array [],
  "ects": null,
  "end_date": "12/31/2100",
  "exercise_completions_needed": null,
  "has_certificate": false,
  "hidden": null,
  "id": Any<String>,
  "inherit_settings_from": null,
  "name": "course1",
  "open_university_registration_links": Array [],
  "order": null,
  "photo": null,
  "points_needed": null,
  "promote": null,
  "slug": "course1",
  "start_date": "01/01/1900",
  "start_point": null,
  "status": "Upcoming",
  "study_module_order": null,
  "study_module_start_point": null,
  "study_modules": Array [
    Object {
      "id": "00000000-0000-0000-0000-000000000101",
    },
    Object {
      "id": "00000000-0000-0000-0000-000000000102",
    },
  ],
  "support_email": null,
  "teacher_in_charge_email": "e@mail.com",
  "teacher_in_charge_name": "teacher",
  "tier": null,
  "upcoming_active_link": false,
  "user_course_settings_visibilities": Array [],
}
`;

exports[`course queries course normal user returns correct language 1`] = `
Object {
  "description": "course1_description_en_US",
  "id": Any<String>,
  "link": "http://link.com",
  "name": "course1_en_US",
  "slug": "course1",
}
`;

exports[`course queries course normal user returns course on id and slug 1`] = `
Object {
  "description": "",
  "id": Any<String>,
  "link": "",
  "name": "course1",
  "slug": "course1",
}
`;

exports[`course queries course normal user returns course on id and slug 2`] = `
Object {
  "description": "",
  "id": Any<String>,
  "link": "",
  "name": "course1",
  "slug": "course1",
}
`;

exports[`course queries courses returns courses 1`] = `
Array [
  Object {
    "course_translations": Array [
      Object {
        "id": "00000000-0000-0000-0000-000000000022",
        "language": "fi_FI",
        "name": "course2_fi_FI",
      },
    ],
    "description": "",
    "hidden": null,
    "id": "00000000-0000-0000-0000-000000000001",
    "link": "",
    "name": "course2",
    "order": null,
    "photo": null,
    "promote": null,
    "slug": "course2",
    "start_point": null,
    "status": "Upcoming",
    "study_module_order": null,
    "study_module_start_point": null,
    "study_modules": Array [
      Object {
        "id": "00000000-0000-0000-0000-000000000102",
        "slug": "module1",
      },
    ],
    "upcoming_active_link": false,
    "user_course_settings_visibilities": Array [],
  },
  Object {
    "course_translations": Array [
      Object {
        "id": "00000000-0000-0000-0000-000000000011",
        "language": "en_US",
        "name": "course1_en_US",
      },
      Object {
        "id": "00000000-0000-0000-0000-000000000012",
        "language": "fi_FI",
        "name": "course1_fi_FI",
      },
    ],
    "description": "",
    "hidden": null,
    "id": "00000000-0000-0000-0000-000000000002",
    "link": "",
    "name": "course1",
    "order": null,
    "photo": null,
    "promote": null,
    "slug": "course1",
    "start_point": null,
    "status": "Upcoming",
    "study_module_order": null,
    "study_module_start_point": null,
    "study_modules": Array [
      Object {
        "id": "00000000-0000-0000-0000-000000000101",
        "slug": "module2",
      },
      Object {
        "id": "00000000-0000-0000-0000-000000000102",
        "slug": "module1",
      },
    ],
    "upcoming_active_link": false,
    "user_course_settings_visibilities": Array [],
  },
]
`;

exports[`course queries courses returns courses filtered by language: courses-bogus 1`] = `Array []`;

exports[`course queries courses returns courses filtered by language: courses-en_US 1`] = `
Array [
  Object {
    "course_translations": Array [
      Object {
        "id": "00000000-0000-0000-0000-000000000011",
        "language": "en_US",
        "name": "course1_en_US",
      },
      Object {
        "id": "00000000-0000-0000-0000-000000000012",
        "language": "fi_FI",
        "name": "course1_fi_FI",
      },
    ],
    "description": "course1_description_en_US",
    "hidden": null,
    "id": "00000000-0000-0000-0000-000000000002",
    "link": "http://link.com",
    "name": "course1_en_US",
    "order": null,
    "photo": null,
    "promote": null,
    "slug": "course1",
    "start_point": null,
    "status": "Upcoming",
    "study_module_order": null,
    "study_module_start_point": null,
    "study_modules": Array [
      Object {
        "id": "00000000-0000-0000-0000-000000000101",
        "slug": "module2",
      },
      Object {
        "id": "00000000-0000-0000-0000-000000000102",
        "slug": "module1",
      },
    ],
    "upcoming_active_link": false,
    "user_course_settings_visibilities": Array [],
  },
]
`;

exports[`course queries courses returns courses filtered by language: courses-fi_FI 1`] = `
Array [
  Object {
    "course_translations": Array [
      Object {
        "id": "00000000-0000-0000-0000-000000000022",
        "language": "fi_FI",
        "name": "course2_fi_FI",
      },
    ],
    "description": "course2_description_fi_FI",
    "hidden": null,
    "id": "00000000-0000-0000-0000-000000000001",
    "link": "",
    "name": "course2_fi_FI",
    "order": null,
    "photo": null,
    "promote": null,
    "slug": "course2",
    "start_point": null,
    "status": "Upcoming",
    "study_module_order": null,
    "study_module_start_point": null,
    "study_modules": Array [
      Object {
        "id": "00000000-0000-0000-0000-000000000102",
        "slug": "module1",
      },
    ],
    "upcoming_active_link": false,
    "user_course_settings_visibilities": Array [],
  },
  Object {
    "course_translations": Array [
      Object {
        "id": "00000000-0000-0000-0000-000000000011",
        "language": "en_US",
        "name": "course1_en_US",
      },
      Object {
        "id": "00000000-0000-0000-0000-000000000012",
        "language": "fi_FI",
        "name": "course1_fi_FI",
      },
    ],
    "description": "course1_description_fi_FI",
    "hidden": null,
    "id": "00000000-0000-0000-0000-000000000002",
    "link": "http:/link.fi.com",
    "name": "course1_fi_FI",
    "order": null,
    "photo": null,
    "promote": null,
    "slug": "course1",
    "start_point": null,
    "status": "Upcoming",
    "study_module_order": null,
    "study_module_start_point": null,
    "study_modules": Array [
      Object {
        "id": "00000000-0000-0000-0000-000000000101",
        "slug": "module2",
      },
      Object {
        "id": "00000000-0000-0000-0000-000000000102",
        "slug": "module1",
      },
    ],
    "upcoming_active_link": false,
    "user_course_settings_visibilities": Array [],
  },
]
`;

exports[`course queries courses returns courses ordered: courses-asc 1`] = `
Array [
  Object {
    "course_translations": Array [
      Object {
        "id": "00000000-0000-0000-0000-000000000011",
        "language": "en_US",
        "name": "course1_en_US",
      },
      Object {
        "id": "00000000-0000-0000-0000-000000000012",
        "language": "fi_FI",
        "name": "course1_fi_FI",
      },
    ],
    "description": "",
    "hidden": null,
    "id": "00000000-0000-0000-0000-000000000002",
    "link": "",
    "name": "course1",
    "order": null,
    "photo": null,
    "promote": null,
    "slug": "course1",
    "start_point": null,
    "status": "Upcoming",
    "study_module_order": null,
    "study_module_start_point": null,
    "study_modules": Array [
      Object {
        "id": "00000000-0000-0000-0000-000000000101",
        "slug": "module2",
      },
      Object {
        "id": "00000000-0000-0000-0000-000000000102",
        "slug": "module1",
      },
    ],
    "upcoming_active_link": false,
    "user_course_settings_visibilities": Array [],
  },
  Object {
    "course_translations": Array [
      Object {
        "id": "00000000-0000-0000-0000-000000000022",
        "language": "fi_FI",
        "name": "course2_fi_FI",
      },
    ],
    "description": "",
    "hidden": null,
    "id": "00000000-0000-0000-0000-000000000001",
    "link": "",
    "name": "course2",
    "order": null,
    "photo": null,
    "promote": null,
    "slug": "course2",
    "start_point": null,
    "status": "Upcoming",
    "study_module_order": null,
    "study_module_start_point": null,
    "study_modules": Array [
      Object {
        "id": "00000000-0000-0000-0000-000000000102",
        "slug": "module1",
      },
    ],
    "upcoming_active_link": false,
    "user_course_settings_visibilities": Array [],
  },
]
`;

exports[`course queries courses returns courses ordered: courses-desc 1`] = `
Array [
  Object {
    "course_translations": Array [
      Object {
        "id": "00000000-0000-0000-0000-000000000022",
        "language": "fi_FI",
        "name": "course2_fi_FI",
      },
    ],
    "description": "",
    "hidden": null,
    "id": "00000000-0000-0000-0000-000000000001",
    "link": "",
    "name": "course2",
    "order": null,
    "photo": null,
    "promote": null,
    "slug": "course2",
    "start_point": null,
    "status": "Upcoming",
    "study_module_order": null,
    "study_module_start_point": null,
    "study_modules": Array [
      Object {
        "id": "00000000-0000-0000-0000-000000000102",
        "slug": "module1",
      },
    ],
    "upcoming_active_link": false,
    "user_course_settings_visibilities": Array [],
  },
  Object {
    "course_translations": Array [
      Object {
        "id": "00000000-0000-0000-0000-000000000011",
        "language": "en_US",
        "name": "course1_en_US",
      },
      Object {
        "id": "00000000-0000-0000-0000-000000000012",
        "language": "fi_FI",
        "name": "course1_fi_FI",
      },
    ],
    "description": "",
    "hidden": null,
    "id": "00000000-0000-0000-0000-000000000002",
    "link": "",
    "name": "course1",
    "order": null,
    "photo": null,
    "promote": null,
    "slug": "course1",
    "start_point": null,
    "status": "Upcoming",
    "study_module_order": null,
    "study_module_start_point": null,
    "study_modules": Array [
      Object {
        "id": "00000000-0000-0000-0000-000000000101",
        "slug": "module2",
      },
      Object {
        "id": "00000000-0000-0000-0000-000000000102",
        "slug": "module1",
      },
    ],
    "upcoming_active_link": false,
    "user_course_settings_visibilities": Array [],
  },
]
`;
